Document free memory pointer overflow check in Sol->Yul. (#11714)
ethereum/solidity: docs/ir/ir-breaking-changes.rst

  - Old code generator: ``aMod = 0`` and ``mMod = 2``
  - New code generator: ``aMod = 4`` and ``mMod = 0``

- The new code generator imposes a hard limit of ``type(uint64).max`` (``0xffffffffffffffff``) for the free memory pointer. Allocations that would increase its value beyond this limit revert. The old code generator does not have this limit.

  For example:

  .. code-block:: solidity

      // SPDX-License-Identifier: GPL-3.0
      pragma solidity >0.8.0;
      contract C {
          function f() public {
              uint[] memory arr;
              // allocation size: 576460752303423481
              // assumes freeMemPtr points to 0x80 initially
              uint solYulMaxAllocationBeforeMemPtrOverflow = (type(uint64).max - 0x80 - 31) / 32;
              // freeMemPtr overflows UINT64_MAX

>               arr = new uint[](solYulMaxAllocationBeforeMemPtrOverflow);

Reviewer: The optimized code for this function is:

```
                    if eq(0x26121ff0, shr(224, calldataload(_1)))
                    {
                        if callvalue() { revert(_1, _1) }
                        if slt(add(calldatasize(), not(3)), _1) { revert(_1, _1) }
                        mstore(_1, shl(224, 0x4e487b71))
                        mstore(4, 0x41)
                        revert(_1, 0x24)
                    }
```

Author: That's nice. Constant propagation I guess. Optimized code would look bigger if length of the dynamic array is a function parameter.

  - Old code generator: ``aMod = 0`` and ``mMod = 2``
  - New code generator: ``aMod = 4`` and ``mMod = 0``

> - The old code generator does not revert when an impending memory allocation exceeds the free memory pointer
> upper bound of ``type(uint64).max`` (``0xffffffffffffffff``).
>   The new code generator reverts in this scenario.

Reviewer 2: ```suggestion
- The new code generator imposes a hard limit of ``type(uint64).max`` (``0xffffffffffffffff``) for the free memory pointer. Allocations that would increase its value beyond this limit revert. The old code generator does not have this limit.
```

  For example:

  .. code-block:: solidity

      // SPDX-License-Identifier: GPL-3.0
      pragma solidity >0.8.0;
      contract C {
          function f() public {
              uint[] memory arr;
              // allocation size: 576460752303423481
              // assumes freeMemPtr points to 0x80 initially
              uint solYulMaxAllocationBeforeMemPtrOverflow = (type(uint64).max - 0x80 - 31) / 32;
              // freeMemPtr overflows UINT64_MAX
              arr = new uint[](solYulMaxAllocationBeforeMemPtrOverflow);
          }
      }

  The function `f()` behaves as follows:

>   - Old code generator: runs out of gas due to the large memory allocation due to array expansion

Reviewer 2: ```suggestion
  - Old code generator: runs out of gas while zeroing the array contents after the large memory allocation
```
